File: btd700/device.py
"""
High-level Sennheiser BTD Dongle Device Controller.
Provides structured commands to inspect and configure the dongle.
"""

import time
import threading
from typing import Optional, Dict, Any, List, Tuple, Callable

from .constants import (
    SENNHEISER_VID,
    BTD600_PID,
    BTD700_PID,
    ReportID,
    HostCmd,
    DongleCmd,
    AudioMode,
    TransportMode,
    SinkMode,
    CodecBit,
    CODEC_NAMES,
    DongleState,
    LEAudioState,
    AudioFrequency,
    AudioResolution,
    BroadcastState,
    BroadcastQuality,
    BroadcastEncryption,
    DeviceType,
    FREQ_LABELS,
    RES_LABELS,
    QUALITY_LABELS,
)
from .protocol import (
    build_host_command,
    parse_dongle_response,
    decode_audio_mode_and_transport,
    decode_supported_codecs,
    decode_codec_in_use,
    decode_dongle_state,
    decode_le_audio_state,
    decode_audio_quality,
    decode_broadcast_info,
    decode_sink_support_transport,
)
from .hid import HidrawDevice, find_sennheiser_hid_devices


class BTDDevice:
    """Represents a connected Sennheiser BTD 600 or BTD 700 dongle."""

    def __init__(self, dev_info: Dict[str, Any]):
        self.dev_info = dev_info
        self.model_name = dev_info.get("model_name", "BTD Dongle")
        self.serial = dev_info.get("serial", "")
        self.vid = dev_info.get("vid", SENNHEISER_VID)
        self.pid = dev_info.get("pid", BTD700_PID)
        self.is_btd700 = (self.pid == BTD700_PID)
        
        # Primary control HID interface (handles Report ID 52 for BTD 700)
        self.control_dev = HidrawDevice(
            dev_info["dev_path"],
            self.vid,
            self.pid,
            name=self.model_name,
            report_ids=dev_info.get("report_ids", [])
        )
        self._lock = threading.Lock()
        self._last_state: Dict[str, Any] = {}

    def open(self) -> bool:
        return self.control_dev.open()

    def close(self):
        self.control_dev.close()

    def __enter__(self):
        self.open()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    def send_command(self, cmd: HostCmd, args: Optional[bytes] = None, wait_resp: bool = True, timeout_ms: int = 1500) -> Optional[Tuple[int, bytes]]:
        """
        Send a BTD 700 Host Command and wait for response.
        """
        with self._lock:
            packet = build_host_command(cmd, args)
            self.control_dev.write(packet)
            
            if not wait_resp:
                return None
            
            start_t = time.time()
            while (time.time() - start_t) * 1000 < timeout_ms:
                raw = self.control_dev.read(64, timeout_ms=300)
                if not raw:
                    continue
                resp = parse_dongle_response(raw)
                if resp:
                    resp_cmd, resp_payload = resp
                    # Verify if response matches command or notification
                    if resp_cmd == int(cmd):
                        return resp
                    # Some commands map to specific response IDs
                    if cmd == HostCmd.GET_AUDIO_MODE_AND_TRANSPORT and resp_cmd == 1:
                        return resp
                    if cmd == HostCmd.GET_SINK_SUPPORT_TRANSPORT and resp_cmd == 21:
                        return resp
            return None

    def get_status(self) -> Dict[str, Any]:
        """Query comprehensive status of the dongle."""
        status: Dict[str, Any] = {
            "model": self.model_name,
            "serial": self.serial,
            "vid": f"0x{self.vid:04X}",
            "pid": f"0x{self.pid:04X}",
            "connected": True,
        }

        if not self.is_btd700:
            # BTD 600 only supports DFU / firmware update via UPM
            return status

        # 1. Dongle State
        resp = self.send_command(HostCmd.GET_DONGLE_STATE)
        if resp:
            state = decode_dongle_state(resp[1])
            status["dongle_state"] = state.name
            status["dongle_state_code"] = state.value

        # 2. Audio Mode & Transport
        resp = self.send_command(HostCmd.GET_AUDIO_MODE_AND_TRANSPORT)
        if resp:
            mode_info = decode_audio_mode_and_transport(resp[1])
            status["audio_mode"] = mode_info["audio_mode"].name if isinstance(mode_info["audio_mode"], AudioMode) else str(mode_info["audio_mode"])
            status["audio_mode_code"] = int(mode_info["audio_mode"])
            status["transport_mode"] = mode_info["transport_mode"].name if isinstance(mode_info["transport_mode"], TransportMode) else str(mode_info["transport_mode"])
            status["transport_mode_code"] = int(mode_info["transport_mode"])
            if "connected_transport" in mode_info:
                ct = mode_info["connected_transport"]
                status["connected_transport"] = ct.name if isinstance(ct, TransportMode) else str(ct)

        # 3. Audio Quality (Sample Rate & Bit Depth)
        resp = self.send_command(HostCmd.GET_AUDIO_QUALITY)
        if resp:
            quality_info = decode_audio_quality(resp[1])
            freq = quality_info["frequency"]
            res = quality_info["resolution"]
            status["frequency"] = FREQ_LABELS.get(freq, "Unknown")
            status["frequency_code"] = freq.value if freq else 0
            status["resolution"] = RES_LABELS.get(res, "Unknown")
            status["resolution_code"] = res.value if res else 0

        # 4. Supported Codecs
        resp = self.send_command(HostCmd.GET_SUPPORTED_CODEC)
        if resp:
            status["supported_codecs"] = decode_supported_codecs(resp[1])

        # 5. Codec in Use
        resp = self.send_command(HostCmd.GET_CODEC_IN_USE)
        if resp:
            in_use = decode_codec_in_use(resp[1])
            status["codec_in_use"] = in_use["name"]
            status["codec_in_use_bit"] = in_use["bit"]
            status["codec_in_use_mask"] = in_use["mask"]

        # 6. LE Audio State
        resp = self.send_command(HostCmd.GET_LE_AUDIO_STATE)
        if resp:
            le_state = decode_le_audio_state(resp[1])
            status["le_audio_state"] = le_state.name
            status["le_audio_state_code"] = le_state.value

        # 7. Auracast Broadcast Info
        resp = self.send_command(HostCmd.GET_BROADCAST_INFO)
        if resp:
            bcast = decode_broadcast_info(resp[1])
            status["broadcast_enabled"] = (bcast["state"] == BroadcastState.ON_PUBLIC)
            status["broadcast_state"] = bcast["state"].name
            status["broadcast_quality"] = QUALITY_LABELS.get(bcast["quality"], bcast["quality"].name)
            status["broadcast_quality_code"] = bcast["quality"].value
            status["broadcast_encrypted"] = (bcast["encryption"] == BroadcastEncryption.ON)

        # 8. Broadcast Name
        resp = self.send_command(HostCmd.GET_BROADCAST_NAME)
        if resp:
            status["broadcast_name"] = resp[1].decode("utf-8", errors="replace").strip("\x00")

        # 9. Broadcast Encryption Key / Password
        resp = self.send_command(HostCmd.GET_BROADCAST_ENCRYPT_KEY)
        if resp:
            status["broadcast_key"] = resp[1].decode("ascii", errors="replace").strip("\x00")

        # 10. Supported Sink Transport
        resp = self.send_command(HostCmd.GET_SINK_SUPPORT_TRANSPORT)
        if resp:
            sink = decode_sink_support_transport(resp[1])
            status["sink_transport_mode"] = sink.name

        self._last_state = status
        return status

    def set_audio_mode(self, mode: AudioMode, transport: Optional[TransportMode] = None) -> bool:
        """
        Switch audio mode (0=One-to-One / High Quality, 1=Gaming, 2=Broadcast).
        """
        if not self.is_btd700:
            return False
        
        current_transport = 1  # Default BT Classic
        if transport is None:
            if "transport_mode_code" in self._last_state:
                current_transport = self._last_state["transport_mode_code"]
        else:
            current_transport = int(transport)

        args = bytes([int(mode), current_transport])
        resp = self.send_command(HostCmd.SET_AUDIO_MODE_AND_TRANSPORT, args=args)
        return resp is not None

    def set_codec(self, codec_bit_idx: int) -> bool:
        """
        Set active codec by bit index (0=SBC, 1=aptX Classic, 2=aptX Adaptive, 3=aptX Lossless, 4=aptX Lite, 5=LC3).
        """
        if not self.is_btd700:
            return False
        mask = 1 << codec_bit_idx
        resp = self.send_command(HostCmd.SET_CODEC_TO_USE, args=bytes([mask]))
        return resp is not None

    def set_broadcast_config(
        self,
        enable: Optional[bool] = None,
        name: Optional[str] = None,
        quality: Optional[BroadcastQuality] = None,
        key: Optional[str] = None
    ) -> bool:
        """Configure Auracast broadcast parameters."""
        if not self.is_btd700:
            return False

        # Set Name
        if name is not None:
            name_bytes = name.encode("utf-8")
            self.send_command(HostCmd.SET_BROADCAST_NAME, args=name_bytes)

        # Set Key
        if key is not None:
            key_bytes = key.encode("ascii")
            self.send_command(HostCmd.SET_BROADCAST_ENCRYPT_KEY, args=key_bytes)

        # Set Broadcast Info (State, Quality, Encryption)
        curr_bcast = self.send_command(HostCmd.GET_BROADCAST_INFO)
        state_val = 0
        qual_val = 2
        enc_val = 0
        if curr_bcast:
            bcast_dec = decode_broadcast_info(curr_bcast[1])
            state_val = int(bcast_dec["state"])
            qual_val = int(bcast_dec["quality"])
            enc_val = int(bcast_dec["encryption"])

        if enable is not None:
            state_val = 1 if enable else 0
        if quality is not None:
            qual_val = int(quality)
        if key is not None:
            enc_val = 1 if len(key) > 0 else 0

        info_bytes = bytes([state_val, qual_val, enc_val])
        resp = self.send_command(HostCmd.SET_BROADCAST_INFO, args=info_bytes)
        return resp is not None

    def trigger_pairing(self) -> bool:
        """Trigger Bluetooth connect / pairing mode."""
        resp = self.send_command(HostCmd.SET_BT_CONNECT, args=b'\x01')
        return resp is not None

    def factory_reset(self) -> bool:
        """Perform a factory reset on the dongle."""
        resp = self.send_command(HostCmd.SET_FACTORY_RESET, args=None)
        return resp is not None


def get_first_dongle() -> Optional[BTDDevice]:
    """Find and return the first connected Sennheiser BTD dongle."""
    devs = find_sennheiser_hid_devices()
    if not devs:
        return None
    return BTDDevice(devs[0])

File: src/device.rs
//! BTD 600 / BTD 700 device communication abstraction.

use std::time::{Duration, Instant};
use serde::{Deserialize, Serialize};

use crate::constants::*;
use crate::hid::{find_sennheiser_hid_devices, DeviceInfo, HidrawDevice};
use crate::protocol::Protocol;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodecInfo {
    pub name: String,
    pub bit: u8,
    pub mask: u32,
    pub is_active: bool,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct DongleStatus {
    pub connected: bool,
    pub model: String,
    pub serial: String,
    pub vid: String,
    pub pid: String,
    pub dongle_state: String,
    pub dongle_state_code: u8,
    pub audio_mode: String,
    pub audio_mode_code: u8,
    pub transport_mode: String,
    pub transport_mode_code: u8,
    pub connected_transport: String,
    pub frequency: String,
    pub resolution: String,
    pub codec_in_use: String,
    pub codec_in_use_bit: Option<u8>,
    pub supported_codecs: Vec<CodecInfo>,
    pub broadcast_enabled: bool,
    pub broadcast_quality: String,
    pub broadcast_quality_code: u8,
    pub broadcast_name: String,
    pub broadcast_encrypted: bool,
}

pub struct BTDDevice {
    pub hid: HidrawDevice,
    pub is_btd700: bool,
    last_status: DongleStatus,
}

impl BTDDevice {
    pub fn new(info: DeviceInfo) -> Self {
        let is_btd700 = info.pid == PID_BTD700;
        Self {
            hid: HidrawDevice::new(info),
            is_btd700,
            last_status: DongleStatus::default(),
        }
    }

    pub fn open(&mut self) -> std::io::Result<()> {
        self.hid.open()
    }

    pub fn close(&mut self) {
        self.hid.close()
    }

    pub fn send_command(
        &mut self,
        cmd: HostCmd,
        args: &[u8],
        timeout_ms: u64,
    ) -> Option<(u8, Vec<u8>)> {
        let packet = Protocol::build_host_cmd(cmd, args);
        if self.hid.write_report(&packet).is_err() {
            return None;
        }

        let start = Instant::now();
        let expected_cmd = cmd as u8;
        let timeout = Duration::from_millis(timeout_ms);

        while start.elapsed() < timeout {
            if let Ok(Some(raw)) = self.hid.read_report(20) {
                if let Some((resp_cmd, payload)) = Protocol::parse_response(&raw) {
                    if resp_cmd == expected_cmd || (expected_cmd == 1 && resp_cmd == 1) || (expected_cmd == 21 && resp_cmd == 21) {
                        return Some((resp_cmd, payload));
                    }
                }
            }
        }

        None
    }

    pub fn get_status(&mut self) -> DongleStatus {
        let mut st = DongleStatus {
            connected: true,
            model: self.hid.info.model_name.clone(),
            serial: self.hid.info.serial.clone(),
            vid: format!("0x{:04X}", self.hid.info.vid),
            pid: format!("0x{:04X}", self.hid.info.pid),
            ..Default::default()
        };

        if !self.is_btd700 {
            self.last_status = st.clone();
            return st;
        }

        const CMD_TIMEOUT: u64 = 80;

        // 1. Dongle State
        if let Some((_, p)) = self.send_command(HostCmd::GetDongleState, &[], CMD_TIMEOUT) {
            if !p.is_empty() {
                st.dongle_state_code = p[0];
                st.dongle_state = dongle_state_str(p[0]).to_string();
            }
        }

        // 2. Audio Mode & Transport
        if let Some((_, p)) = self.send_command(HostCmd::GetAudioModeAndTransport, &[], CMD_TIMEOUT) {
            if p.len() >= 2 {
                let am = AudioMode::from_u8(p[0]);
                let tm = TransportMode::from_u8(p[1]);
                st.audio_mode_code = p[0];
                st.audio_mode = am.as_str().to_string();
                st.transport_mode_code = p[1];
                st.transport_mode = tm.as_str().to_string();
                if p.len() >= 3 {
                    st.connected_transport = TransportMode::from_u8(p[2]).as_str().to_string();
                } else {
                    st.connected_transport = st.transport_mode.clone();
                }
            }
        }

        // 3. Audio Quality
        if let Some((_, p)) = self.send_command(HostCmd::GetAudioQuality, &[], CMD_TIMEOUT) {
            if p.len() >= 2 {
                st.frequency = audio_frequency_str(p[0]).to_string();
                st.resolution = audio_resolution_str(p[1]).to_string();
            }
        }

        // 4. Codec in use
        let mut active_bit = None;
        if let Some((_, p)) = self.send_command(HostCmd::GetCodecInUse, &[], CMD_TIMEOUT) {
            if !p.is_empty() {
                let mask = p[0];
                for b in 0..6 {
                    if (mask & (1 << b)) != 0 {
                        active_bit = Some(b);
                        if let Some(cb) = CodecBit::from_bit(b) {
                            st.codec_in_use = cb.name().to_string();
                            st.codec_in_use_bit = Some(b);
                        }
                        break;
                    }
                }
            }
        }

        // 5. Supported Codecs
        if let Some((_, p)) = self.send_command(HostCmd::GetSupportedCodec, &[], CMD_TIMEOUT) {
            if !p.is_empty() {
                let mask = p[0];
                let mut list = Vec::new();
                for b in 0..6 {
                    if (mask & (1 << b)) != 0 {
                        if let Some(cb) = CodecBit::from_bit(b) {
                            list.push(CodecInfo {
                                name: cb.name().to_string(),
                                bit: b,
                                mask: 1 << b,
                                is_active: Some(b) == active_bit,
                            });
                        }
                    }
                }
                st.supported_codecs = list;
            }
        }

        // 6. Broadcast Info
        if let Some((_, p)) = self.send_command(HostCmd::GetBroadcastInfo, &[], CMD_TIMEOUT) {
            if p.len() >= 3 {
                st.broadcast_enabled = p[0] == 1;
                st.broadcast_quality_code = p[1];
                st.broadcast_quality = BroadcastQuality::from_u8(p[1]).as_str().to_string();
                st.broadcast_encrypted = p[2] == 1;
            }
        }

        // 7. Broadcast Name
        if let Some((_, p)) = self.send_command(HostCmd::GetBroadcastName, &[], CMD_TIMEOUT) {
            st.broadcast_name = String::from_utf8_lossy(&p).to_string();
        }

        self.last_status = st.clone();
        st
    }

    pub fn set_audio_mode(&mut self, mode: AudioMode) -> bool {
        if !self.is_btd700 {
            return false;
        }
        let transport = self.last_status.transport_mode_code.max(1);
        let args = [mode as u8, transport];
        self.send_command(HostCmd::SetAudioModeAndTransport, &args, 1000).is_some()
    }

    pub fn set_codec(&mut self, codec_bit: u8) -> bool {
        if !self.is_btd700 {
            return false;
        }
        let mask = 1u8 << codec_bit;
        self.send_command(HostCmd::SetCodecToUse, &[mask], 1000).is_some()
    }

    pub fn set_broadcast_config(
        &mut self,
        enable: Option<bool>,
        name: Option<&str>,
        quality: Option<BroadcastQuality>,
        key: Option<&str>,
    ) -> bool {
        if !self.is_btd700 {
            return false;
        }

        if let Some(n) = name {
            self.send_command(HostCmd::SetBroadcastName, n.as_bytes(), 800);
        }

        if let Some(k) = key {
            self.send_command(HostCmd::SetBroadcastEncryptKey, k.as_bytes(), 800);
        }

        let curr_info = self.send_command(HostCmd::GetBroadcastInfo, &[], 500);
        let mut st_val = 0u8;
        let mut qual_val = 2u8;
        let mut enc_val = 0u8;

        if let Some((_, p)) = curr_info {
            if p.len() >= 3 {
                st_val = p[0];
                qual_val = p[1];
                enc_val = p[2];
            }
        }

        if let Some(en) = enable {
            st_val = if en { 1 } else { 0 };
        }
        if let Some(q) = quality {
            qual_val = q as u8;
        }
        if let Some(k) = key {
            enc_val = if !k.is_empty() { 1 } else { 0 };
        }

        let info = [st_val, qual_val, enc_val];
        self.send_command(HostCmd::SetBroadcastInfo, &info, 1000).is_some()
    }

    pub fn trigger_pairing(&mut self) -> bool {
        self.send_command(HostCmd::SetBTConnect, &[1], 1000).is_some()
    }

    pub fn factory_reset(&mut self) -> bool {
        self.send_command(HostCmd::SetFactoryReset, &[], 1000).is_some()
    }
}

pub fn get_first_dongle() -> Option<BTDDevice> {
    let devs = find_sennheiser_hid_devices();
    if devs.is_empty() {
        return None;
    }
    Some(BTDDevice::new(devs[0].clone()))
}
